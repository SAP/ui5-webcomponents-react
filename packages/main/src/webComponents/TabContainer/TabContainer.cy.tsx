import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';
import { Tab } from '../Tab';
import { TabContainer } from './index';

describe('TabContainer', () => {
  it('Basic Test (generated)', () => {
    cy.mount(
      <TabContainer>
        <Tab text={'Tab 1'} />
      </TabContainer>
    );
    cy.get('[ui5-tabcontainer]').should('exist');
  });

  cypressPassThroughTestsFactory(TabContainer, { children: <Tab text={'Tab 1'} /> });
});
