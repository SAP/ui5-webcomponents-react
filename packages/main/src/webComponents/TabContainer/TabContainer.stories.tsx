import type { Meta, StoryObj } from '@storybook/react';
import activitiesIcon from '@ui5/webcomponents-icons/dist/activities.js';
import addIcon from '@ui5/webcomponents-icons/dist/add.js';
import employeeIcon from '@ui5/webcomponents-icons/dist/employee.js';
import menuIcon from '@ui5/webcomponents-icons/dist/menu.js';
import settingsIcon from '@ui5/webcomponents-icons/dist/settings.js';
import { TabContainerBackgroundDesign, TabLayout } from '../../enums/index.js';
import { Tab } from '../Tab/index.js';
import { TabSeparator } from '../TabSeparator/index.js';
import { TabContainer } from './index.js';

const meta = {
  title: 'Layouts & Floorplans / TabContainer',
  component: TabContainer,
  argTypes: {
    children: { control: { disable: true } },
    overflowButton: { control: { disable: true } }
  },
  args: {
    tabLayout: TabLayout.Standard,
    contentBackgroundDesign: TabContainerBackgroundDesign.Solid,
    headerBackgroundDesign: TabContainerBackgroundDesign.Solid
  }
} satisfies Meta<typeof TabContainer>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: (args) => {
    return (
      <TabContainer {...args}>
        <Tab text="Tab One" icon={menuIcon} selected additionalText="5">
          Content Tab 1
        </Tab>
        <Tab text="Tab Two" icon={activitiesIcon} additionalText="20">
          Content Tab 2
        </Tab>
        <Tab text="Tab Three" icon={addIcon}>
          Content Tab 3
        </Tab>
        <Tab text="Tab Four" icon={employeeIcon}>
          Content Tab 4
        </Tab>
        <Tab text="Tab Five" icon={settingsIcon}>
          Content Tab 5
        </Tab>
      </TabContainer>
    );
  }
};

export const WithTabSeparator: Story = {
  name: 'with TabSeparator',
  render: (args) => {
    return (
      <TabContainer {...args}>
        <Tab text="Tab One" icon={menuIcon} selected additionalText="5">
          Content Tab 1
        </Tab>
        <TabSeparator />
        <Tab text="Tab Two" icon={activitiesIcon} additionalText="20">
          Content Tab 2
        </Tab>
      </TabContainer>
    );
  }
};

export const WithNestedTabs: Story = {
  name: 'with nested tabs',
  render: (args) => {
    return (
      <TabContainer {...args}>
        <Tab
          text="Tab One"
          selected
          subTabs={
            <>
              <Tab text="Tab 1.1" />
              <Tab text="Tab 1.2" />
              <Tab text="Tab 1.3" />
            </>
          }
        >
          <div style={{ display: 'none' }} />
        </Tab>
        <TabSeparator />
        <Tab text="Tab Two" />
        <Tab
          text="Tab Three"
          subTabs={
            <>
              <Tab text="Tab 3.1" />
              <Tab text="Tab 3.2" />
              <Tab text="Tab 3.3" />
            </>
          }
        />
      </TabContainer>
    );
  }
};
