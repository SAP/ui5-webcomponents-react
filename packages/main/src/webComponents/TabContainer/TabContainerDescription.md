## Structure

The `TabContainer` can hold two types of entities:

*   `Tab` - contains all the information on an item (text and icon)
*   `TabSeparator` - used to separate tabs with a line

## Hierarchies

Multiple sub tabs could be placed underneath one main tab. Nesting allows deeper hierarchies with indentations to indicate the level of each nested tab. When a tab has both sub tabs and own content its click area is split to allow the user to display the content or alternatively to expand / collapse the list of sub tabs.

## CSS Shadow Parts

<ui5-link target="_blank" href="https://developer.mozilla.org/en-US/docs/Web/CSS/::part">CSS Shadow Parts</ui5-link> allow developers to style elements inside the Shadow DOM.  
The `TabContainer` exposes the following CSS Shadow Parts:

*   content - Used to style the content of the component

## Keyboard Handling

### Fast Navigation

This component provides a build in fast navigation group which can be used via `F6 / Shift + F6` or `Ctrl + Alt(Option) + Down / Ctrl + Alt(Option) + Up`. In order to use this functionality, you need to import the following module: `import "@ui5/webcomponents-base/dist/features/F6Navigation.js"`