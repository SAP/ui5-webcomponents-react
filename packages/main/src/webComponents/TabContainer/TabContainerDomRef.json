[
  {
    "name": "allItems",
    "readonly": "true",
    "visibility": "public",
    "type": "undefined",
    "description": "Returns all slotted tabs and their subTabs in a flattened array. The order of tabs is depth-first. For example, given the following slotted elements: <code> <ui5-tabcontainer> <ui5-tab id=\"First\" text=\"First\"> ... <ui5-tab slot=\"subTabs\" id=\"Nested\" text=\"Nested\">...</ui5-tab> </ui5-tab> <ui5-tab id=\"Second\" text=\"Second\">...</ui5-tab> <ui5-tab-separator id=\"sep\"></ui5-tab-separator> <ui5-tab id=\"Third\" text=\"Third\">...</ui5-tab> </ui5-tabcontainer> </code> Calling <code>allItems</code> on this TabContainer will return the instances in the following order: <code>[ ui5-tab#First, ui5-tab#Nested, ui5-tab#Second, ui5-tab-separator#sep, ui5-tab#Third ]</code>"
  }
]
