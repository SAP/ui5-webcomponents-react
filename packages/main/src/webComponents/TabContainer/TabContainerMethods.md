## Attributes & Methods

This component exposes public attributes and methods. You can use them directly on the instance of the component, e.g. by using React Refs.

| Name                    | Parameters | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
| ----------------------- | ---------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **allItems (readonly)** | &mdash;    | Returns all slotted tabs and their subTabs in a flattened array. The order of tabs is depth-first. For example, given the following slotted elements: <code> <ui5-tabcontainer> <ui5-tab id="First"> ... <ui5-tab slot="subTabs" id="Nested">...</ui5-tab> </ui5-tab> <ui5-tab id="Second">...</ui5-tab> <ui5-tab-separator id="sep"></ui5-tab-separator> <ui5-tab id="Third">...</ui5-tab> </ui5-tabcontainer> </code> Calling <code>allItems</code> on this TabContainer will return the instances in the following order: <code>[ ui5-tab#First, ui5-tab#Nested, ui5-tab#Second, ui5-tab-separator#sep, ui5-tab#Third ]</code> |
