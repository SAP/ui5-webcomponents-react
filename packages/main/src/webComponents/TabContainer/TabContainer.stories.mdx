import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { TabContainer } from '@ui5/webcomponents-react/dist/TabContainer';
import { TabLayout } from '@ui5/webcomponents-react/dist/TabLayout';
import { Tab } from '@ui5/webcomponents-react/dist/Tab';
import { SemanticColor } from '@ui5/webcomponents-react/dist/SemanticColor';
import { TabSeparator } from '@ui5/webcomponents-react/dist/TabSeparator';
import { DocsHeader } from '@docs/DocsHeader';
import { Footer } from '@docs/Footer';
import '@ui5/webcomponents-icons/dist/menu.js';
import '@ui5/webcomponents-icons/dist/activities.js';
import '@ui5/webcomponents-icons/dist/add.js';
import '@ui5/webcomponents-icons/dist/calendar.js';
import '@ui5/webcomponents-icons/dist/settings.js';
import '@ui5/webcomponents-icons/dist/sys-monitor.js';

<Meta
  title="Layouts & Floorplans / TabContainer"
  component={TabContainer}
  subcomponents={{ Tab, TabSeparator }}
  argTypes={{
    children: { control: { disable: true } },
    overflowButton: { control: { disable: true } }
  }}
  args={{
    tabLayout: TabLayout.Standard
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return (
        <TabContainer {...args}>
          <Tab text="Tab One" icon="menu" selected additionalText="5">
            Content Tab 1
          </Tab>
          <Tab text="Tab Two" icon="activities" additionalText="20">
            Content Tab 2
          </Tab>
          <Tab text="Tab Three" icon="add">
            Content Tab 3
          </Tab>
          <Tab text="Tab Four" icon="employee">
            Content Tab 4
          </Tab>
          <Tab text="Tab Five" icon="settings">
            Content Tab 5
          </Tab>
        </TabContainer>
      );
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />

## Structure

The `TabContainer` can hold two types of entities:

- `Tab` - contains all the information on an item (text and icon)
- `TabSeparator` - used to separate tabs with a vertical line

<br />

# More Examples

<br />

## TabContainer with customizable Tab

The <code>Tab</code> represents a selectable item inside a <code>TabContainer</code>. It defines both the item in the
tab strip (top part of the <code>TabContainer</code>) and the content that is presented to the user once the tab is
selected

<Canvas>
  <Story
    name="with customizable Tab"
    args={{
      additionalText: '10',
      disabled: false,
      icon: 'sys-monitor',
      selected: true,
      design: SemanticColor.Positive,
      text: 'Customizable Tab',
      children: 'Tab Content'
    }}
    argTypes={{
      additionalText: { description: `Represents the "additionalText" text, which is displayed in the tab filter.` },
      disabled: { description: `Enabled items can be selected.` },
      icon: {
        description: `Defines the icon source URI to be displayed as graphical element within the <code>Tab</code>. The SAP-icons font provides numerous built-in icons. See all the available icons in the <ui5-link target="_blank" href="https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html" class="api-table-content-cell-link">Icon Explorer</ui5-link>.`
      },
      selected: { description: `Specifies if the <code>Tab</code> is selected.` },
      design: {
        options: Object.values(SemanticColor),
        control: { type: 'radio' }
      },
      stableDomRef: { description: `Defines the stable selector that you can use via getStableDomRef method.` },
      text: { description: `The text to be displayed for the item.` },
      children: { description: `Defines the tab content.`, control: { disable: false } },
      collapsed: { table: { disable: true } },
      fixed: { table: { disable: true } },
      showOverflow: { table: { disable: true } },
      tabLayout: { table: { disable: true } },
      overflowButton: { table: { disable: true } },
      onTabSelect: { table: { disable: true } }
    }}
  >
    {(args) => {
      return (
        <TabContainer {...args}>
          <Tab {...args} />
          <Tab text="Tab 2" icon="activities">
            Content Tab 2
          </Tab>
          <Tab text="Tab 3" icon="add">
            Content Tab 3
          </Tab>
          <Tab text="Tab 4" icon="employee">
            Content Tab 4
          </Tab>
          <Tab text="Tab 5" icon="settings">
            Content Tab 5
          </Tab>
        </TabContainer>
      );
    }}
  </Story>
</Canvas>

<ArgsTable story="with customizable Tab" />

<br />

## TabContainer with TabSeparator

The <code>TabSeparator</code> represents a vertical line to separate tabs inside a <code>TabContainer</code>

<Canvas>
  <Story name="with TabSeparator">
    {(args) => {
      return (
        <TabContainer {...args}>
          <Tab text="Tab One" icon="menu" selected additionalText="5">
            Content Tab 1
          </Tab>
          <TabSeparator />
          <Tab text="Tab Two" icon="activities" additionalText="20">
            Content Tab 2
          </Tab>
        </TabContainer>
      );
    }}
  </Story>
</Canvas>

<Footer />
