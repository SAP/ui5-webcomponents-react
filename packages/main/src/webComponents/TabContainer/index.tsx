import { TabLayout } from '@ui5/webcomponents-react/dist/TabLayout';
import { withWebComponent } from '@ui5/webcomponents-react/dist/withWebComponent';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import { Ui5CustomEvent } from '@ui5/webcomponents-react/interfaces/Ui5CustomEvent';
import { Ui5DomRef } from '@ui5/webcomponents-react/interfaces/Ui5DomRef';
import { ReactNode } from 'react';

import '@ui5/webcomponents/dist/TabContainer.js';

interface TabContainerAttributes {
  /**
   * Defines whether the tab content is collapsed.
   */
  collapsed?: boolean;
  /**
   * Defines whether the tabs are in a fixed state that is not expandable/collapsible by user interaction.
   */
  fixed?: boolean;
  /**
   * Defines whether the overflow select list is displayed.
   *
   * The overflow select list represents a list, where all tab filters are displayed so that it's easier for the user to select a specific tab filter.
   */
  showOverflow?: boolean;
  /**
   * Defines the alignment of the content and the `additionalText` of a tab.
   *
   * **Note:** The content and the `additionalText` would be displayed vertically by defualt, but when set to `Inline`, they would be displayed horizontally.
   *
   * Available options are:
   *
   * *   `Standard`
   * *   `Inline`
   */
  tabLayout?: TabLayout | keyof typeof TabLayout;
}

export interface TabContainerDomRef extends TabContainerAttributes, Ui5DomRef {}

export interface TabContainerPropTypes extends TabContainerAttributes, CommonProps {
  /**
   * Defines the tabs.
   *
   * **Note:** Use `Tab` and `TabSeparator` for the intended design.
   */
  children?: ReactNode | ReactNode[];
  /**
   * Defines the button which will open the overflow menu. If nothing is provided to this slot, the default button will be used.
   *
   * __Note:__ When passing a custom React component to this prop, you have to make sure your component reads the `slot` prop and appends it to the most outer element of your component.
   * Learn more about it [here](https://sap.github.io/ui5-webcomponents-react/?path=/docs/knowledge-base--page#adding-custom-components-to-slots).
   */
  overflowButton?: ReactNode;
  /**
   * Fired when a tab is selected.
   */
  onTabSelect?: (event: Ui5CustomEvent<HTMLElement, { tab: ReactNode; tabIndex: number }>) => void;
}

/**
 * The `TabContainer` represents a collection of tabs with associated content. Navigation through the tabs changes the content display of the currently active content area. A tab can be labeled with text only, or icons with text.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/TabContainer" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const TabContainer = withWebComponent<TabContainerPropTypes, TabContainerDomRef>(
  'ui5-tabcontainer',
  ['tabLayout'],
  ['collapsed', 'fixed', 'showOverflow'],
  ['overflowButton'],
  ['tab-select']
);

TabContainer.displayName = 'TabContainer';

TabContainer.defaultProps = {
  tabLayout: TabLayout.Standard
};

export { TabContainer };
