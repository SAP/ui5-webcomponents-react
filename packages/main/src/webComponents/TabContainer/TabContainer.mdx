import { ArgTypesWithNote, ControlsWithNote, DocsHeader, DomRefTable, Footer } from '@sb/components';
import SubcomponentsSection from '@sb/docs/SubcomponentsSection.md?raw';
import { Canvas, Markdown, Meta } from '@storybook/blocks';
import { Tab } from '../Tab';
import TabDescription from '../Tab/TabDescription.md?raw';
import { TabSeparator } from '../TabSeparator';
import TabSeparatorDescription from '../TabSeparator/TabSeparatorDescription.md?raw';
import * as ComponentStories from './TabContainer.stories';
import Description from './TabContainerDescription.md?raw';
import TabContainerDomRef from './TabContainerDomRef.json';
import { excludePropsForAbstract } from '@sb/utils';

<Meta of={ComponentStories} />

<DocsHeader />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Markdown>{Description}</Markdown>

<DomRefTable rows={TabContainerDomRef} />

# More Examples

## TabContainer with TabSeparator

The <code>TabSeparator</code> represents a vertical line to separate tabs inside a <code>TabContainer</code>

<Canvas of={ComponentStories.WithTabSeparator} />

## TabContainer with nested Tabs

<Canvas of={ComponentStories.WithNestedTabs} />

<Markdown>{SubcomponentsSection}</Markdown>

## Tab

<Markdown>{TabDescription}</Markdown>
<ArgTypesWithNote hideHTMLPropsNote exclude={excludePropsForAbstract} of={Tab} />

## TabSeparator

<Markdown>{TabSeparatorDescription}</Markdown>
<ArgTypesWithNote hideHTMLPropsNote exclude={excludePropsForAbstract} of={TabSeparator} />

<Footer />
