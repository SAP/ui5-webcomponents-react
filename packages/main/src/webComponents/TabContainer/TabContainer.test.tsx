import { render } from '@shared/tests';
import { createCustomPropsTest } from '@shared/tests/utils';
import { TabContainer } from '@ui5/webcomponents-react/dist/TabContainer';
import React from 'react';

describe('TabContainer', () => {
  test('Basic Test (generated)', () => {
    const { asFragment } = render(<TabContainer />);
    expect(asFragment()).toMatchSnapshot();
  });
  createCustomPropsTest(TabContainer);
});
