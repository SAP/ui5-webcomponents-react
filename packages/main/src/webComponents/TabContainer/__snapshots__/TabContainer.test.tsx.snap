// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabContainer Basic Test (generated) 1`] = `
<DocumentFragment>
  <ui5-tabcontainer
    data-sap-ui-fastnavgroup="true"
    tab-layout="Standard"
    tabs-overflow-mode="End"
    ui5-tabcontainer=""
  />
</DocumentFragment>
`;
